modules/frontend/frontend.go

```diff
@@ -0,0 +1,227 @@
+package frontend
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"github.com/cortexproject/cortex/pkg/querier/frontend"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/user"
+	"net/http"
+)
+
+type FrontendConfig struct {
+	frontend.Config `yaml:",inline"`
+	ShardNum int    `yaml:"shard_num,omitempty"`
+}
+
+func (cfg *FrontendConfig) ApplyDefaults() {
+	cfg.Config.CompressResponses = false
+	cfg.Config.DownstreamURL = ""
+	cfg.Config.LogQueriesLongerThan = 0
+	cfg.Config.MaxOutstandingPerTenant = 100
+	cfg.ShardNum = 4
+}
+
+// NewTripperware returns a Tripperware configured with a middleware to split requests
+func NewTripperware(cfg FrontendConfig, logger log.Logger, registerer prometheus.Registerer) (frontend.Tripperware, error) {
+	level.Info(logger).Log("msg", "creating tripperware in query frontend to shard queries")
+	queriesPerTenant := promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "query_frontend_queries_total",
+		Help:      "Total queries sent per tenant.",
+	}, []string{"user"})
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		// get the http request, add some custom parameters to it, split it, and call downstream roundtripper
+		rt := NewRoundTripper(next, ShardingWare(cfg.ShardNum, logger, registerer))
+		return frontend.RoundTripFunc(func(r *http.Request) (*http.Response, error) {
+			level.Info(util.Logger).Log("msg", "request received by custom tripperware")
+			orgID := r.Header.Get(user.OrgIDHeaderName)
+			queriesPerTenant.WithLabelValues(orgID).Inc()
+
+			r = r.WithContext(user.InjectOrgID(r.Context(), orgID))
+			return rt.RoundTrip(r)
+		})
+	}, nil
+}
+
+type Handler interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type Middleware interface {
+	Wrap(Handler) Handler
+}
+
+// MiddlewareFunc is like http.HandlerFunc, but for Middleware.
+type MiddlewareFunc func(Handler) Handler
+
+// Wrap implements Middleware.
+func (q MiddlewareFunc) Wrap(h Handler) Handler {
+	return q(h)
+}
+
+func MergeMiddlewares(middleware ...Middleware) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		for i := len(middleware) - 1; i >= 0; i-- {
+			next = middleware[i].Wrap(next)
+		}
+		return next
+	})
+}
+
+type roundTripper struct {
+	next    http.RoundTripper
+	handler Handler
+}
+
+// NewRoundTripper merges a set of middlewares into an handler, then inject it into the `next` roundtripper
+func NewRoundTripper(next http.RoundTripper, middlewares ...Middleware) http.RoundTripper {
```

Comment:
this seems largely copy/paste from Cortex?  Why are we bringing all this code in?  It seems like the only truly custom code is the `shardQuery.Do()` method